doc: add instructions on how to test unstable builds on macOS (#924)
jellyfin/jellyfin.org: docs/general/testing/macos.md

---
uid: testing-macos
title: macOS
description: Test unstable builds on macOS.
sidebar_position: 2
---

# Test unstable builds on macOS

> Testing unstable builds requires you to be confident with the command line, or at least not afraid of it.

Reviewer: ```suggestion
Testing unstable builds requires you to not be afraid of using the command line.
```

## Get Unstable Server

> Unlike the release version, the unstable builds for macOS will not come with a packaged DMG or application. Instead, you have to run them from the command line.

Reviewer: ```suggestion
Unlike the stable releases, the unstable builds for macOS do not come with a packaged installer or application. Instead, they have to be ran from the command line.
```

The latest unstable builds for packaged jellyfin can be obtained here:

- [Intel Macs](https://repo.jellyfin.org/?path=/server/macos/latest-unstable/amd64)
- [Apple Silicon Macs](https://repo.jellyfin.org/?path=/server/macos/latest-unstable/arm64)

> There is no difference in content between the `tar.gz` and the `tar.xz` archives; they are only compressed in different formats. Either one will work.

Reviewer: ```suggestion
There is no difference in content between the `tar.gz` and the `tar.xz` archives. They are only compressed in different formats and either one will work.
```

## Get Jellyfin FFmpeg

> The unstable version doesn't come with FFmpeg by default, so you need to obtain FFmpeg manually.

Reviewer: ```suggestion
The unstable builds don't come with FFmpeg bundled. Instead, the binaries have to be downloaded separately.
```

Jellyfin's custom FFmpeg is available [here](https://repo.jellyfin.org/?path=/ffmpeg/macos).

Select the version corresponding to the unstable version of the server.

Intel Macs should download the `AMD64` binaries.

> Apple Silicon Macs should download the `ARM64` binaries.

Reviewer: I see the difference between AMD64 and ARM64 mentioned twice. Maybe it would be cleaner to mention it once on the top?

Author: This is actually the first time it appears as human-readable form. The former one is in a link and the user will just see "Intel" and "Apple Silicon"

Reviewer: What I meant was put the Apple Silicon -> ARM64, Intel -> AMD64 on top and make both the upper level links

## Prepare The Unstable Server

> After you have downloaded both the server and FFmpeg, you should extract them in a folder you like.
> 
> Before running, we need to remove quarantine flag and do an ad-hoc sign to the server binaries:

Reviewer: ```suggestion
After downloading both the server and FFmpeg, extract them to a folder you like. Before running, the quarantine flag needs to be removed and an ad-hoc signature needs to be created for the binaries.
```

> Open a command line interface. You can use the `Terminal.app` come with macOS, then run following commands:
> 
> ```shell
> cd /path/to/jellyfin
> xattr -rd com.apple.quarantine .
> codesign -fs - --deep jellyfin
> ```

Reviewer: ```suggestion
Open a terminal in the directory containing the extracted contents of the server package. This can be done by right-clicking on the directory and selecting `open a terminal here`. Run the following commands in the terminal:
```shell
xattr -rd com.apple.quarantine .
codesign -fs - --deep jellyfin
```
```

I don't actually know the exact string for that option since my system language isn't English but you get the idea.

Author: There will be not "Open a terminal here" option on standard macOS install.

Reviewer: Oh sorry I forgot that I have an automator script installed for that

## Prepare The FFmpeg

We also need to perform the similar thing to our ffmpeg binaries, but this time we only have to remove the quarantine flag as the ad-hoc signature will be automatically created on first launch.

Open another command line interface, then run the following commands:

```shell
cd /path/to/jellyfin-ffmpeg
xattr -rd com.apple.quarantine .

> ```

Reviewer: If this section and the section before it are doing the same thing, can they be combined into one section?

## Run The Unstable Server

> At this point, you should be able to run unstable jellyfin with the specified ffmpeg in the command line:

Reviewer: ```suggestion
At this point, the binaries should be ready for testing. Run the following command in a terminal to start Jellyfin:
```
There should probably be a reminder to backup any existing Jellyfin installs in case of the need to downgrade to a previous Jellyfin version

```shell
/path/to/jellyfin-folder/jellyfin --ffmpeg /path/to/ffmpeg-folder/ffmpeg

> ```

Reviewer: Won't it cause problem if the user doesn't first cd to /path/to/jellyfin ?

Author: we are using absolute path to the jellyfin binary, so this won't be a issue.

> After pressing enter, you should see Jellyfin up and running, with logs appearing in the console.

Reviewer: ```suggestion
After running the command, Jellyfin should startup in the terminal window.
```

> To stop it, you press `CTRL+C` in the terminal window.

Reviewer: ```suggestion
To stop Jellyfin, press `CONTROL+C` in the terminal window, and Jellyfin will shutdown gracefully.
```
the CTRL key is labeled `control` on mac keyboards.
